language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
         self.locals.insert(idx, value);
     }
 
-    /// Return true if self may safely be destroyed
-    pub fn is_safe_to_destroy(&self) -> bool {
-        self.locals.values().all(|x| x.is_safe_to_destroy())
-            && self.globals.values().all(|x| x.is_empty())
+    /// checks if a local may be safely destroyed
+    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
+        match self.locals[&idx] {
+            AbstractValue::Reference(_) => true,
+            AbstractValue::Value(Kind::All) | AbstractValue::Value(Kind::Resource) => false,
+            AbstractValue::Value(Kind::Unrestricted) => !self.is_local_borrowed(idx),
+        }
     }
 
-    /// Return true if local@idx may safely be destroyed
-    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
-        self.local(idx).is_safe_to_destroy()
+    /// checks if the stack frame of the function being analyzed can be safely destroyed
+    pub fn is_frame_safe_to_destroy(&self) -> bool {
+        self.locals.values().all(|x| x.is_unrestricted_value())
+            && !self.is_nonce_borrowed(self.frame_root)
     }
 
     /// destroys local@idx
-    /// call only if self.is_local_safe_to_destroy(idx) returns true
     pub fn destroy_local(&mut self, idx: LocalIndex) {
+        checked_verify!(self.is_local_safe_to_destroy(idx));
         let local = self.locals.remove(&idx).unwrap();
         match local {
             AbstractValue::Reference(nonce) => self.destroy_nonce(nonce),
-            AbstractValue::Value(kind, borrowed_nonces) => {
-                checked_verify!(kind == Kind::Unrestricted && borrowed_nonces.is_empty());
+            AbstractValue::Value(kind) => {
+                checked_verify!(kind == Kind::Unrestricted);
             }
         }
     }
 
-    /// nonce must be fresh
+    /// adds nonce to borrow graph
+    /// nonce must be freshly allocated
```

Comment:
Is there a situation where a nonce is not freshly allocated? If so, is nonce an appropriate name? Is there a way this precondition can be enforced?

I also find myself wondering if Nonce can be cloned or copied? If not, then the move semantics of Rust will guarantee that this parameter will always be a unique nonce.

Reply:
I personally like Todd's naming of this (`RefID`). It made it a lot more obvious to me under that naming scheme.

Reply:
The mysterious requirement of freshness is the following precondition:
checked_precondition!(self.0[&to].len() == 0);

I have deleted the documentation line containing the word fresh to avoid confusion.

Nonce is simply a wrapped number.  There are invariants about the borrow graph being maintained but a nonce by itself does not satisfy any particular property.

Reply:
My use of the word "nonce" has caused a lot of confusion.  I am happy to rename.  @hermanventer: what do you think of RefID, the suggestion by @tzakian?  

Also, is it ok to do the renaming in a separate diff?

Reply:
Nonce -> RefID renaming will be done in a separate PR.